### PR TITLE
Azure labels and tags support for old refresh mech

### PR DESCRIPTION
**Pull request by AlexanderZagaynov**
https://bugzilla.redhat.com/show_bug.cgi?id=1568806

new refresh mech: https://github.com/ManageIQ/manageiq-providers-azure/pull/229
related core repo PR: https://github.com/ManageIQ/manageiq/pull/17212
related UI repo PR: https://github.com/ManageIQ/manageiq-ui-classic/pull/3697

### CODE REVIEW

File: app/models/manageiq/providers/azure/cloud_manager/refresh_parser.rb [hunk 1]
```diff
@@ -266,6 +268,8 @@ def parse_instance(instance)
 
         rg_ems_ref = get_resource_group_ems_ref(instance)
 
+        labels = parse_labels(instance.tags)
```
Comment (cben) [inline app/models/manageiq/providers/azure/cloud_manager/refresh_parser.rb]: These are new right, you didn't collect them before?
Terminology note:
in container provider we called them "labels" in code and used section `labels` because [that's what Kubernetes calls them](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/).
IIUC in AWS and Azure they original concept is called tags?
It gets confusing if we collect "tags" into "labels" and map those into MIQ "tags" :-(

I propose code following UI.  Are you going to display azure tags in UI, when looking at a VM?  How would you call them?
Even if you don't display them, IIRC mapping add/edit UI currently speaks of "label name", "label values" and that needs to be rethought.

- IMHO "single pane of glass" consistency is not a relevant consideration here, if "label" won't mean anything to a user using MIQ with only Azure.
- I can see 2 things called "tags" being confusing as well, but IMHO "Azure tags" vs "ManageIQ/Cloudforms tags" (or "Smart Management tags" if we must™, ugh) could work.

Comment (AlexanderZagaynov) [reply app/models/manageiq/providers/azure/cloud_manager/refresh_parser.rb]: I believe it is better to stay in line with Amazon's code patterns here to ease further DRYing (I'm not going to DRY it right now, because it should be backported).
Maybe with the https://github.com/ManageIQ/manageiq-providers-azure/pull/229 PR (which is not suggested for backporting), or with a fully separate one.
What about the UI: I'm not sure how it should be presented there, so I'm aligning to the Amazon's side again for simplicity.

**Comment by djberg96** [comment]
@Ladas You mentioned on the Amazon PR that the specs wouldn't pass until it was also implemented for graph refresh. I assume that's true here as well, correct?

**Comment by Ladas** [comment]
@djberg96 yes, the specs are testing both old refresh and graph refresh, so we either need to add a condition or implement it in both

**Comment by JPrause** [comment]
@miq-bot add_label blocker

**Comment by AlexanderZagaynov** [comment]
I'm closing this in favor of #229 because of https://github.com/ManageIQ/manageiq-providers-azure/pull/231#issuecomment-378499055

**Comment by AlexanderZagaynov** [comment]
https://travis-ci.org/ManageIQ/manageiq-providers-azure/jobs/364148489#L1938
failing because of https://github.com/ManageIQ/manageiq/pull/17212 not yet merged

https://travis-ci.org/ManageIQ/manageiq-providers-azure/jobs/364148489#L1947
failing because of https://github.com/ManageIQ/manageiq-providers-azure/pull/229/commits/04d3085b46ab29325b73cdd5d47b08720daa2c62 can't be merged (for backporting compatibility)

@Ladas @djberg96 @bronaghs any ideas how to workaround the second issue?

**Comment by bronaghs** [comment]
@AlexanderZagaynov Why can't you update the `vm.custom_attributes.size` and `vm.tags.count `from 0 to 1, why cant that be backported to the Gaprindishvili branch?

**Comment by AlexanderZagaynov** [comment]
@bronaghs other tests will fail (that code piece used for all refresh types)

**Comment by bronaghs** [comment]
sounds like those tests needs updated too then if the count legitimately increased on the provider, no?

File: spec/models/manageiq/providers/azure/cloud_manager/refresh_parser_spec.rb [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+  context 'labels and tags' do
+    let(:vm) do
+      VCR.use_cassette(described_class.name.underscore,
+                       :allow_unused_http_interactions => true) { EmsRefresh.refresh(ems) }
+      ems.vms.find_by(:name => 'az-test')
+    end
+
+    it 'fetches labels' do
+      expect(vm.labels.map { |l| [l.name, l.value] }.to_h).to eq('redhat' => 'brno', 'test' => 'hohoho')
```
Comment (bronaghs) [inline spec/models/manageiq/providers/azure/cloud_manager/refresh_parser_spec.rb]: @AlexanderZagaynov - thinking of readability, can this be simplified to:

`let (:cust_attr) do
    vm.labels.first
end`

expect(cust_attr.name).to eq('redhat')
expect(cust_attr.value).to eq('brno')

**Comment by bronaghs** [comment]
hi @AlexanderZagaynov - I am curious why you created a new spec and didnt just add these tests to an existing file? The first 23ish lines are duplicates.

File: spec/models/manageiq/providers/azure/cloud_manager/refresher_spec.rb [hunk 3]
```diff
@@ -118,8 +122,8 @@
           assert_specific_flavor
           assert_specific_disk
           assert_specific_security_group
-          assert_specific_vm_powered_on
-          assert_specific_vm_powered_off
+          assert_specific_vm_powered_on(**assertion_options)
+          assert_specific_vm_powered_off(**assertion_options)
           assert_specific_template
```
Comment (bronaghs) [inline spec/models/manageiq/providers/azure/cloud_manager/refresher_spec.rb]: do you need the double splat when you are passing in 1 argument?

**Comment by bronaghs** [comment]
Looks good @AlexanderZagaynov

**Comment by AlexanderZagaynov** [comment]
@miq-bot add_label "fine/no"

**Comment by AlexanderZagaynov** [comment]
@miq-bot add_label fine/no

**Comment by JPrause** [comment]
@AlexanderZagaynov can you add the gaprindashvili/yes and the fine/yes if this can be backported.

**Comment by AlexanderZagaynov** [comment]
@bronaghs specs are not failing now

File: spec/models/manageiq/providers/azure/cloud_manager/refresher_spec.rb [hunk 4]
```diff
@@ -108,6 +113,9 @@
       end
 
       it "will perform a full refresh" do
+        # NOTE: for backporting compatibility, where tagging does not implemented for the Graph refresh
```
Comment (martinpovolny) [inline spec/models/manageiq/providers/azure/cloud_manager/refresher_spec.rb]: The comment does not sound English ;-)
I am not sure how to fix it. Please, ask some native speaker.

Comment (JPrause) [reply spec/models/manageiq/providers/azure/cloud_manager/refresher_spec.rb]: Not sure of the context but I'd suggest either "is not implemented" or "can not be implemented".

**Comment by JPrause** [comment]
@AlexanderZagaynov please add the gaprindashvili/yes label if this can be backported. Can you also verify the fine/no label.

File: spec/models/manageiq/providers/azure/cloud_manager/refresher_spec.rb (hunk 4 above)
Comment (cben) [reply spec/models/manageiq/providers/azure/cloud_manager/refresher_spec.rb]: Perhaps "where tag mapping was not implemented"

Comment (Ladas) [reply spec/models/manageiq/providers/azure/cloud_manager/refresher_spec.rb]: lets just fix this and then I think its ready

File: spec/models/manageiq/providers/azure/cloud_manager/refresher_spec.rb [hunk 5]
```diff
@@ -31,6 +31,11 @@
         @managed_data_disk = "miqazure-linux-managed-data-disk"
         @template          = nil
         @avail_zone        = nil
+
+        FactoryGirl.create(:tag_mapping_with_category,
+                           :labeled_resource_type => 'VmAzure',
```
Comment (Ladas) [inline spec/models/manageiq/providers/azure/cloud_manager/refresher_spec.rb]: the class should be ``` ManageIQ::Providers::Azure::CloudManager::Vm```

File: app/models/manageiq/providers/azure/cloud_manager/refresh_parser.rb [hunk 6]
```diff
@@ -279,6 +283,8 @@ def parse_instance(instance)
           :orchestration_stack => @data_index.fetch_path(:orchestration_stacks, @resource_to_stack[uid]),
           :availability_zone   => @data_index.fetch_path(:availability_zones, 'default'),
           :resource_group      => @data_index.fetch_path(:resource_groups, rg_ems_ref),
+          :labels              => labels,
+          :tags                => map_labels('VmAzure', labels),
```
Comment (Ladas) [inline app/models/manageiq/providers/azure/cloud_manager/refresh_parser.rb]: the class should be ``` ManageIQ::Providers::Azure::CloudManager::Vm```

Comment (AlexanderZagaynov) [reply app/models/manageiq/providers/azure/cloud_manager/refresh_parser.rb]: that is handled there: https://github.com/ManageIQ/manageiq-ui-classic/pull/3697/files#diff-95b50e997ace14a4e385c9a176eea25fR85
and we discussed it with @cben a lot, and agreed on current implementation, please don't make us starting that over again :)

File: spec/models/manageiq/providers/azure/cloud_manager/refresher_spec.rb (hunk 4 above)
Comment (AlexanderZagaynov) [reply spec/models/manageiq/providers/azure/cloud_manager/refresher_spec.rb]: done

File: app/models/manageiq/providers/azure/cloud_manager/refresh_parser.rb (hunk 6 above)
Comment (cben) [reply app/models/manageiq/providers/azure/cloud_manager/refresh_parser.rb]: @Ladas this is not supposed to be a class name.  it's just a string that matches what UI puts in ContainerLabelTagMappings (if constrained to one type).  you can see a list of currently used values in Alexander's link ^^ in *keys* of MAPPABLE_ENTITIES const.  see for example `"Image"` which is not a class name.

Yes, this arrangement could be improved as a whole (with a migration — out of scope here), but as it is, adding 1 fully qualified model name to that set would only confuse matters IMO.

**Comment by cben** [comment]
locally, cherry-picks to gaprindashvili cleanly, and passes tests.
@miq-bot add-label gaprindashvili/yes, fine/conflict

**Comment by AlexanderZagaynov** [comment]
There is a separate PR for **Fine**: https://github.com/ManageIQ/manageiq-providers-azure/pull/242

**Comment by AlexanderZagaynov** [comment]
@miq-bot remove-label fine/conflict

**Comment by simaishi** [comment]
Backported to Fine via https://github.com/ManageIQ/manageiq-providers-azure/pull/242

**Comment by simaishi** [comment]
Gaprindashvili backport details:

```
$ git log -1
commit ea578dbf0e2eb76dc34aca04ba18fb2fd59684ce
Author: Bronagh Sorota <bsorota@redhat.com>
Date:   Wed Apr 18 13:48:32 2018 +0200

    Merge pull request #231 from AlexanderZagaynov/BZ-1493041-old-refresh
    
    Azure labels and tags support for old refresh mech
    (cherry picked from commit 3014aee644b6280207332f7ac9a108b4c618b800)
    
    https://bugzilla.redhat.com/show_bug.cgi?id=1568807
```
